### PR TITLE
Add support for `meta.url` to enable links in terminal

### LINKED ISSUE #5842: Add support for meta.url to enable links in terminal
### What is the problem you're trying to solve?

As discussed with #2622, I think it is very useful if stylelint rules (including plugins) could have metadata with doc URL, deprecated, etc. [like ESLint](https://eslint.org/docs/developer-guide/working-with-rules#rule-basics).

For example, use cases:

- Tools or plugins like [stylelint-find-rules](https://github.com/alexilyaev/stylelint-find-rules)
- Create rule links on a terminal (see #5835)
- Generate a simple document for rules
- and so on...

Actually, we had metadata with the `deprecated` property:

https://github.com/stylelint/stylelint/blob/16b4595496f54fa52182fafe9d75e78691b38261/lib/rules/at-rule-blacklist/index.js#L62

Also, the Stylelint document below has described rule metadata:

https://github.com/stylelint/stylelint/blob/9e75552d5da4f4448208a65b0c6f876ef9a0feb0/docs/developer-guide/rules.md#L198

### What solution would you like to see?

I propose to define a metadata schema via the document and type definitions (TypeScript).

For example of a rule with metadata:

```js
// lib/rules/color-no-invalid-hex/index.js
rule.ruleName = ruleName;
rule.messages = messages;

rule.meta = {
  type: 'problem', // or 'convention' etc.
  docs: {
    description: 'Disallow invalid hex colors.',
    url: 'https://stylelint.io/user-guide/rules/list/color-no-invalid-hex',
  },
  deprecated: false,
  fixable: true,
  schema: [
    { type: 'boolean' }, // primary
  ],
};

module.exports = rule;
```

For the `Rule` type of the TypeScript definitions, we would add a `meta` property:

https://github.com/stylelint/stylelint/blob/9e75552d5da4f4448208a65b0c6f876ef9a0feb0/types/stylelint/index.d.ts#L154-L164

### PR DESCRIPTION
**Pull request by ybiquitous**
> Which issue, if any, is this issue related to?

Closes #5842
Related to https://github.com/stylelint/stylelint/pull/5835#discussion_r785246463

> Is there anything in the PR that needs further explanation?

This pull request adds the `meta.url` property to all built-in rules. For automation, I use a temporary script as below:

Run:

```console
$ cd stylelint
$ node add-meta-to-stylelint-rules.js
...
$ npm run format
```

Script:

```js
// add-meta-to-stylelint-rules.js
const fs = require('fs');
const globby = require('globby');

const files = globby.sync('lib/rules/*/index.js');

for (const file of files) {
	const content = fs.readFileSync(file).toString();

	const rule = file.replace(/lib\/rules\/([^/]+)\/index.js/, '$1');

	let newContent = content.replace(
		'module.exports = rule;',
		'rule.meta = meta;\nmodule.exports = rule;',
	);

	newContent = newContent.replace(
		/(const messages = ruleMessages\(.+?\);)/s,
		`$1

const meta = {
	url: 'https://stylelint.io/user-guide/rules/list/${rule}',
};`,
	);

	fs.writeFileSync(file, newContent);

	console.log(`${file} (${rule}) updated`); // eslint-disable-line no-console
}
```

### CODE REVIEW

File: types/stylelint/index.d.ts [hunk 1]
```diff
@@ -160,6 +160,7 @@ declare module 'stylelint' {
 		export type Rule<P = any, S = any> = RuleBase<P, S> & {
 			ruleName: string;
 			messages: RuleMessages;
+			meta: { url: string; deprecated?: boolean };
```
Comment (ybiquitous) [inline types/stylelint/index.d.ts:163]: [ask] Since the next version, `rule.meta` would be required. Is this too strict for plugin authors?

Comment (jeddy3) [reply types/stylelint/index.d.ts:163]: I think it should be opt-in, but recommended in the docs.

Comment (ybiquitous) [reply types/stylelint/index.d.ts]: @jeddy3 Thanks for the quick feedback. Marked as optional. (6afacb30)

**Review by jeddy3 — APPROVED** [review]
LGTM, thanks.

(Nice script to automate it!)

**Comment by jeddy3** [comment]
@ybiquitous When you have a mo, can you resolve the conflicts and then we'll merge this and https://github.com/stylelint/stylelint/pull/5845.

**Comment by ybiquitous** [comment]
@jeddy3 Conflicts fixed!

**Review by jeddy3 — APPROVED** [review]
LGTM, thanks!

**Comment by ybiquitous** [comment]
@jeddy3 Don't we need to add this change to `CHANGELOG.md`?

**Comment by Reda011** [comment]
Hello, does  Plugins support meta.url ?

I developed as Document(https://stylelint.io/developer-guide/plugins) said, but it doesn't work. Such as:

eg: I changed something from 'stylelint-z-index-value-constraint'(https://github.com/kristerkari/stylelint-z-index-value-constraint)

`
const stylelint = require("stylelint");

const ruleName = "plugin/z-index-value-constraint";

const messages = stylelint.utils.ruleMessages(ruleName, {
  largerThanMax: expected =>
    `Expected z-index to have maximum value of ${expected}.`,
  smallerThanMin: expected =>
    `Expected z-index to have minimum value of ${expected}.`
});

const meta = {
	url: 'https://www.google.com/',    // !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!  Here it is !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
};


function isNumber(value) {
  return typeof value === "number";
}

function isNegative(value) {
  return value < 0;
}

const _isNaN =
  Number.isNaN ||
  function (value) {
    return value !== value;
  };

function possibleValueTest(value) {
  return isNumber(value) && !isNegative(value);
}

module.exports = stylelint.createPlugin(
  ruleName,
  function (options, secondary) {
    return function (cssRoot, result) {
      const validOptions = stylelint.utils.validateOptions(
        result,
        ruleName,
        {
          actual: options,
          possible: {
            min: possibleValueTest,
            max: possibleValueTest
          }
        },
        {
          actual: secondary,
          possible: {
            ignoreValues: [isNumber]
          },
          optional: true
        }
      );

      if (!validOptions) {
        return;
      }

      cssRoot.walkDecls("z-index", function (decl) {
        const value = Number(decl.value);

        if (
          _isNaN(value) ||
          (secondary &&
            Array.isArray(secondary.ignoreValues) &&
            secondary.ignoreValues.indexOf(value) > -1)
        ) {
          return;
        }

        if (options.max && Math.abs(value) > options.max) {
          stylelint.utils.report({
            ruleName,
            result,
            node: decl,
            message: messages.largerThanMax(
              isNegative(value) ? options.max * -1 : options.max
            )
          });
        }

        if (options.min && Math.abs(value) < options.min) {
          stylelint.utils.report({
            ruleName,
            result,
            node: decl,
            message: messages.smallerThanMin(
              isNegative(value) ? options.min * -1 : options.min
            )
          });
        }
      });
    };
  }
);

module.exports.ruleName = ruleName;
module.exports.messages = messages;
module.exports.meta = meta;

`

**Comment by ybiquitous** [comment]
@Reda011 Thanks for the comment. Maybe, it's a bug. I'll fix it soon, but I appreciate it if you would open an issue.

**Comment by Reda011** [comment]
@ybiquitous Ok, I have opened an issue：https://github.com/stylelint/stylelint/issues/6245

**Pull request merged**